### PR TITLE
Target Index Settings if create index during rollup

### LINKED ISSUE #1376: [FEATURE] Create target index with specific settings while rollup
**Is your feature request related to a problem?**
It is not possible to setup settings for  index created while rollup.

**What solution would you like?**
If during the execution of rollup job index will be created, it will be created with settings passed by user into job settings as follows:
```
...
    "target_index": "rollup-test-index",
    "target_index_settings": {
      "index.codec": "zlib"
    },
...
```
`target_index_settings` can be null. In this case behaviour will be same as before. 

**What alternatives have you considered?**
Only one alternative way is to create index before rollup.

### PR DESCRIPTION
**Pull request by MrChaos1993**
### Description
Add optional field `target_index_settings` to rollup job.

### Related Issues
Resolves #1376 

### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md). [opensearch-project/opensearch-api-specification/pull/822]
- [x] Commits are signed per the DCO using `--signoff`.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose). [opensearch-project/documentation-website/issues/9269]

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/index-management/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by MrChaos1993** [comment]
Hi!
Could someone review this PR?

File: src/main/kotlin/org/opensearch/indexmanagement/rollup/model/Rollup.kt [hunk 1]
```diff
@@ -301,6 +321,10 @@ data class Rollup(
                     DESCRIPTION_FIELD -> description = xcp.text()
                     SOURCE_INDEX_FIELD -> sourceIndex = xcp.text()
                     TARGET_INDEX_FIELD -> targetIndex = xcp.text()
+                    TARGET_INDEX_SETTINGS_FIELD -> {
```
Comment (bowenlan-amzn) [inline src/main/kotlin/org/opensearch/indexmanagement/rollup/model/Rollup.kt:324]: Add UT to cover this

File: src/main/kotlin/org/opensearch/indexmanagement/rollup/model/Rollup.kt [hunk 2]
```diff
@@ -129,6 +135,11 @@ data class Rollup(
         description = sin.readString(),
         sourceIndex = sin.readString(),
         targetIndex = sin.readString(),
+        targetIndexSettings = if (sin.readBoolean()) {
+            Settings.readSettingsFromStream(sin)
+        } else {
+            null
+        },
```
Comment (bowenlan-amzn) [inline src/main/kotlin/org/opensearch/indexmanagement/rollup/model/Rollup.kt:142]: Considering a rolling upgrade cluster, node on old version won't have this logic to understand targetIndexSettings.

https://github.com/opensearch-project/OpenSearch/blob/84477373f70247774f1bd52ad708c393da7f7d7c/libs/core/src/main/java/org/opensearch/Version.java#L119

Stream will have the version, depending on it, we can decide whether to read or write.

Here's an example

https://github.com/opensearch-project/OpenSearch/blob/e3a6ccadc942c64e83bd224031bc4d1c6ab14623/server/src/main/java/org/opensearch/search/aggregations/metrics/CardinalityAggregationBuilder.java#L119-L121

File: src/main/kotlin/org/opensearch/indexmanagement/rollup/model/Rollup.kt [hunk 3]
```diff
@@ -200,6 +216,8 @@ data class Rollup(
         out.writeString(description)
         out.writeString(sourceIndex)
         out.writeString(targetIndex)
+        out.writeBoolean(targetIndexSettings != null)
+        if (targetIndexSettings != null) Settings.writeSettingsToStream(targetIndexSettings, out)
```
Comment (bowenlan-amzn) [inline src/main/kotlin/org/opensearch/indexmanagement/rollup/model/Rollup.kt:220]: Similar ask for Stream

File: src/test/kotlin/org/opensearch/indexmanagement/rollup/model/RollupTests.kt [hunk 4]
```diff
@@ -52,6 +54,16 @@ class RollupTests : OpenSearchTestCase() {
...
     }
 
+    fun `test rollup requires correct target index settings`() {
+        assertFailsWith(SettingsException::class, "Unknown property was `index.codec1`") {
+            randomRollup().copy(targetIndexSettings = Settings.builder().put("index.codec1", "zlib").build())
+        }
+
+        val sb = Settings.builder()
+        sb.put("index.codec", "zlib")
+        randomRollup().copy(targetIndexSettings = sb.build())
```
Comment (bowenlan-amzn) [inline src/test/kotlin/org/opensearch/indexmanagement/rollup/model/RollupTests.kt:64]: Separate to a different test. 
Add test to cover multiple index settings.

File: src/test/kotlin/org/opensearch/indexmanagement/rollup/TestHelpers.kt [hunk 5]
```diff
@@ -112,6 +112,7 @@ fun randomRollup(): Rollup {
         description = OpenSearchRestTestCase.randomAlphaOfLength(10),
         sourceIndex = OpenSearchRestTestCase.randomAlphaOfLength(10).lowercase(Locale.ROOT),
         targetIndex = OpenSearchRestTestCase.randomAlphaOfLength(10).lowercase(Locale.ROOT),
+        targetIndexSettings = null,
```
Comment (bowenlan-amzn) [inline src/test/kotlin/org/opensearch/indexmanagement/rollup/TestHelpers.kt:115]: Maybe sth like
```
if (OpenSearchRestTestCase.randomBoolean()) null else randomSettings,
```

File: src/main/kotlin/org/opensearch/indexmanagement/rollup/model/Rollup.kt (hunk 3 above)
Comment (bowenlan-amzn) [reply src/main/kotlin/org/opensearch/indexmanagement/rollup/model/Rollup.kt:220]: Add UT to cover the round trip, write the object to StreamOutput and read in. There should be some existing tests doing that.

File: src/test/kotlin/org/opensearch/indexmanagement/indexstatemanagement/action/RollupActionIT.kt [hunk 6]
```diff
@@ -42,6 +42,7 @@ class RollupActionIT : IndexStateManagementRestTestCase() {
             ISMRollup(
                 description = "basic search test",
                 targetIndex = "target_rollup_search",
+                targetIndexSettings = null,
```
Comment (bowenlan-amzn) [inline src/test/kotlin/org/opensearch/indexmanagement/indexstatemanagement/action/RollupActionIT.kt:45]: Add one test for targetIndexSettings not null

File: src/test/kotlin/org/opensearch/indexmanagement/rollup/runner/RollupRunnerIT.kt [hunk 7]
```diff
@@ -96,6 +96,7 @@ class RollupRunnerIT : RollupRestTestCase() {
                 description = "basic stats test",
                 sourceIndex = sourceIdxTestName,
                 targetIndex = targetIdxTestName,
+                targetIndexSettings = null,
```
Comment (bowenlan-amzn) [inline src/test/kotlin/org/opensearch/indexmanagement/rollup/runner/RollupRunnerIT.kt:99]: Add one test for targetIndexSettings not null, verify it's correctly applied

File: src/main/kotlin/org/opensearch/indexmanagement/rollup/model/ISMRollup.kt [hunk 8]
```diff
@@ -129,6 +143,8 @@ data class ISMRollup(
     override fun writeTo(out: StreamOutput) {
         out.writeString(description)
         out.writeString(targetIndex)
+        out.writeBoolean(targetIndexSettings != null)
+        if (targetIndexSettings != null) Settings.writeSettingsToStream(targetIndexSettings, out)
```
Comment (bowenlan-amzn) [inline src/main/kotlin/org/opensearch/indexmanagement/rollup/model/ISMRollup.kt:147]: Similar ask for Stream

File: src/main/kotlin/org/opensearch/indexmanagement/rollup/model/ISMRollup.kt [hunk 9]
```diff
@@ -88,6 +96,11 @@ data class ISMRollup(
     constructor(sin: StreamInput) : this(
         description = sin.readString(),
         targetIndex = sin.readString(),
+        targetIndexSettings = if (sin.readBoolean()) {
+            Settings.readSettingsFromStream(sin)
+        } else {
+            null
+        },
```
Comment (bowenlan-amzn) [inline src/main/kotlin/org/opensearch/indexmanagement/rollup/model/ISMRollup.kt:103]: Similar ask for Stream

**Review by bowenlan-amzn — COMMENTED** [review]
Thanks for the contribution!
I suppose on the frontend, we also need some change for this. Please consider contribute or just create an issue there.

File: src/main/kotlin/org/opensearch/indexmanagement/rollup/model/Rollup.kt (hunk 1 above)
Comment (MrChaos1993) [reply src/main/kotlin/org/opensearch/indexmanagement/rollup/model/Rollup.kt:327]: Add UTs to `kotlin/org/opensearch/indexmanagement/rollup/model/XContentTests.kt`

File: src/main/kotlin/org/opensearch/indexmanagement/rollup/model/Rollup.kt (hunk 2 above)
Comment (MrChaos1993) [reply src/main/kotlin/org/opensearch/indexmanagement/rollup/model/Rollup.kt]: Thank you so much for this suggestion! 
I tried to lookup for backward compatibility solution, but couldn't find anything in plugin code.

File: src/main/kotlin/org/opensearch/indexmanagement/rollup/model/Rollup.kt (hunk 3 above)
Comment (MrChaos1993) [reply src/main/kotlin/org/opensearch/indexmanagement/rollup/model/Rollup.kt]: Done

File: src/test/kotlin/org/opensearch/indexmanagement/rollup/model/RollupTests.kt (hunk 4 above)
Comment (MrChaos1993) [reply src/test/kotlin/org/opensearch/indexmanagement/rollup/model/RollupTests.kt]: Done

File: src/test/kotlin/org/opensearch/indexmanagement/rollup/TestHelpers.kt (hunk 5 above)
Comment (MrChaos1993) [reply src/test/kotlin/org/opensearch/indexmanagement/rollup/TestHelpers.kt]: Done

File: src/test/kotlin/org/opensearch/indexmanagement/indexstatemanagement/action/RollupActionIT.kt (hunk 6 above)
Comment (MrChaos1993) [reply src/test/kotlin/org/opensearch/indexmanagement/indexstatemanagement/action/RollupActionIT.kt:48]: Done

File: src/test/kotlin/org/opensearch/indexmanagement/rollup/runner/RollupRunnerIT.kt (hunk 7 above)
Comment (MrChaos1993) [reply src/test/kotlin/org/opensearch/indexmanagement/rollup/runner/RollupRunnerIT.kt:164]: Done

File: src/main/kotlin/org/opensearch/indexmanagement/rollup/model/ISMRollup.kt (hunk 8 above)
Comment (MrChaos1993) [reply src/main/kotlin/org/opensearch/indexmanagement/rollup/model/ISMRollup.kt]: Done

File: src/main/kotlin/org/opensearch/indexmanagement/rollup/model/ISMRollup.kt (hunk 9 above)
Comment (MrChaos1993) [reply src/main/kotlin/org/opensearch/indexmanagement/rollup/model/ISMRollup.kt]: Done

**Comment by MrChaos1993** [comment]
> Thanks for the contribution! I suppose on the frontend, we also need some change for this. Please consider contribute or just create an issue there.

Sure, here is the issue https://github.com/opensearch-project/index-management-dashboards-plugin/issues/1275. I'll try to implement it during this week.

File: src/test/kotlin/org/opensearch/indexmanagement/rollup/TestHelpers.kt [hunk 10]
```diff
@@ -33,8 +37,9 @@ import org.opensearch.indexmanagement.rollup.model.metric.Metric
 import org.opensearch.indexmanagement.rollup.model.metric.Min
 import org.opensearch.indexmanagement.rollup.model.metric.Sum
 import org.opensearch.indexmanagement.rollup.model.metric.ValueCount
+import org.opensearch.test.OpenSearchTestCase
 import org.opensearch.test.rest.OpenSearchRestTestCase
-import java.util.Locale
+import java.util.*
```
Comment (bowenlan-amzn) [inline src/test/kotlin/org/opensearch/indexmanagement/rollup/TestHelpers.kt:42]:
```suggestion
import java.util.Locale
```

**Review by bowenlan-amzn — APPROVED** [review]
